### ISSUE TITLE
[java] AssertionError "this should be unreachable" with scala library

### ISSUE THREAD

**Issue by Laurie-Enveil** [pmd/pmd#5283]
**Affects PMD Version:** 7.3.0, 7.4.0, 7.5.0, 7.60

Make sure, to test with the latest PMD version.

**Description:**

We use a _lot_ of Spark. Our PMD setup is with the [maven pmd plugin](https://github.com/apache/maven-pmd-plugin). We are attempting to upgrade from PMD plugin 3.21.2 which uses PMD v 6.55.0. At that version we are able to run `mvn pmd:pmd` (or `mvn clean install` successfully.  

Upon upgrading to PMD plugin 3.25.0 using PMD 7.3.0, we began encountering errors when scanning files that use Scala. I have also tested with PMD 7.4.0, 7.5.0, and 7.6.0 by [configuring the PMD version at runtime](https://maven.apache.org/plugins/maven-pmd-plugin/examples/upgrading-PMD-at-runtime.html)

**Exception Stacktrace:**

```
Exception in thread "main" java.lang.AssertionError: This should be unreachable: private field of an interface? scala/PartialFunction$AndThen, modifiers: public private static
	at net.sourceforge.pmd.util.AssertionUtil.shouldNotReachHere(AssertionUtil.java:223)
	at net.sourceforge.pmd.util.AssertionUtil.shouldNotReachHere(AssertionUtil.java:216)
	at net.sourceforge.pmd.lang.java.symbols.table.internal.JavaResolvers.isAccessibleIn(JavaResolvers.java:482)
	at net.sourceforge.pmd.lang.java.symbols.table.internal.JavaResolvers.lambda$hidingWalkResolvers$3(JavaResolvers.java:383)
	at net.sourceforge.pmd.lang.java.symbols.table.internal.JavaResolvers.processDeclarations(JavaResolvers.java:435)
	at net.sourceforge.pmd.lang.java.symbols.table.internal.JavaResolvers.walkSelf(JavaResolvers.java:412)
	at net.sourceforge.pmd.lang.java.symbols.table.internal.JavaResolvers.walkSelf(JavaResolvers.java:417)
	at net.sourceforge.pmd.lang.java.symbols.table.internal.JavaResolvers.hidingWalkResolvers(JavaResolvers.java:386)
	at net.sourceforge.pmd.lang.java.symbols.table.internal.JavaResolvers.inheritedMembersResolvers(JavaResolvers.java:358)
	at net.sourceforge.pmd.lang.java.symbols.table.internal.SymTableFactory.typeBody(SymTableFactory.java:395)
	at net.sourceforge.pmd.lang.java.symbols.table.internal.SymbolTableResolver$MyVisitor.visitTypeDecl(SymbolTableResolver.java:283)
	at net.sourceforge.pmd.lang.java.symbols.table.internal.SymbolTableResolver$MyVisitor.visit(SymbolTableResolver.java:319)
	at net.sourceforge.pmd.lang.java.symbols.table.internal.SymbolTableResolver$MyVisitor.visit(SymbolTableResolver.java:138)
	at net.sourceforge.pmd.lang.java.ast.ASTAnonymousClassDeclaration.acceptVisitor(ASTAnonymousClassDeclaration.java:63)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.acceptVisitor(AbstractJavaNode.java:38)
	at net.sourceforge.pmd.lang.java.symbols.table.internal.SymbolTableResolver$MyVisitor.traverse(SymbolTableResolver.java:164)
	at net.sourceforge.pmd.lang.java.symbols.table.internal.SymbolTableResolver.traverse(SymbolTableResolver.java:101)
	at net.sourceforge.pmd.lang.java.internal.JavaAstProcessor.lambda$process$1(JavaAstProcessor.java:132)
	at net.sourceforge.pmd.benchmark.TimeTracker.bench(TimeTracker.java:163)
	at net.sourceforge.pmd.lang.java.internal.JavaAstProcessor.process(JavaAstProcessor.java:132)
	at net.sourceforge.pmd.lang.java.internal.JavaAstProcessor.process(JavaAstProcessor.java:166)
	at net.sourceforge.pmd.lang.java.internal.JavaAstProcessor.process(JavaAstProcessor.java:150)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.parseImpl(JavaParser.java:69)
	at net.sourceforge.pmd.lang.java.ast.JavaParser.parseImpl(JavaParser.java:25)
	at net.sourceforge.pmd.lang.ast.impl.javacc.JjtreeParserAdapter.parse(JjtreeParserAdapter.java:36)
	at net.sourceforge.pmd.lang.impl.PmdRunnable.parse(PmdRunnable.java:112)
	at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource(PmdRunnable.java:132)
	at net.sourceforge.pmd.lang.impl.PmdRunnable.run(PmdRunnable.java:80)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

**Code Sample demonstrating the issue:**

```java
package org.example;

import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.List;

import scala.Option;
import scala.PartialFunction;
import scala.collection.JavaConverters;
import org.apache.spark.sql.catalyst.expressions.Expression;
import org.apache.spark.sql.catalyst.expressions.NamedExpression;
import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression;
import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan;
import org.apache.spark.sql.catalyst.plans.logical.Aggregate;
import scala.runtime.AbstractPartialFunction;

public class Util {

  public static Set<String> getClientPlanFields(LogicalPlan rootPlan) {
    return getSelectedExpressionsFromLogicalPlan(rootPlan).stream()
        .filter(NamedExpression.class::isInstance)
        .map(exp -> ((NamedExpression) exp).name())
        .collect(Collectors.toSet());
  }

  public static List<Expression> getSelectedExpressionsFromLogicalPlan(LogicalPlan rootPlan) {
    Option<Aggregate> maybeAggregate = rootPlan.collectFirst(partial(
        Aggregate.class::isInstance, Aggregate.class::cast));

    List<NamedExpression> aggregateExpressions = JavaConverters.seqAsJavaList(
        maybeAggregate.get().aggregateExpressions());
    return aggregateExpressions.stream().map(Expression.class::cast).collect(Collectors.toList());
  }

  public static <T, R> PartialFunction<T, R> partial(Function<T, Boolean> filter, Function<T, R> applier) {
    return new AbstractPartialFunction<T, R>() {
      @Override
      public boolean isDefinedAt(T item) {
        return filter.apply(item);
      }

      @Override
      public R apply(T item) {
        return applier.apply(item);
      }
    };
  }
}

```

**Steps to reproduce:**

Please provide detailed steps for how we can reproduce the bug.

1.  `mvn clean pmd:pmd`

**Running PMD through:** *[Maven]*

Target JDK 1.8

**Comment by Laurie-Enveil** [comment]
You are right that it is java code when we're using scala. E.g. the `LogicalPlan` and `NamedExpression` and `PartialFunction` are written in scala. 

We are not using any scala rules.

Here's my pom.xml
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <groupId>org.example</groupId>
  <artifactId>pmd-test</artifactId>
  <version>1.0-SNAPSHOT</version>

  <properties>
    <maven.compiler.source>1.8</maven.compiler.source>
    <maven.compiler.target>1.8</maven.compiler.target>
    <maven.javac.target>1.8</maven.javac.target>

    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>

    <dependency.spark.version>3.3.4</dependency.spark.version>
    <dependency.scala.version.suffix>2.12</dependency.scala.version.suffix>
    <pmdVersion>7.6.0</pmdVersion>

  </properties>
  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>org.apache.spark</groupId>
        <artifactId>spark-core_${dependency.scala.version.suffix}</artifactId>
        <version>${dependency.spark.version}</version>
        <exclusions>
          <exclusion>
            <groupId>org.apache.logging.log4j</groupId>
            <artifactId>log4j-1.2-api</artifactId>
          </exclusion>
          <exclusion>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-log4j12</artifactId>
          </exclusion>
          <exclusion>
            <groupId>org.scala-lang</groupId>
            <artifactId>scalap</artifactId>
          </exclusion>
          <exclusion>
            <groupId>org.scala-lang</groupId>
            <artifactId>scala-library</artifactId>
          </exclusion>
          <exclusion>
            <groupId>org.scala-lang</groupId>
            <artifactId>scala-reflect</artifactId>
          </exclusion>
          <exclusion>
            <groupId>org.tukaani</groupId>
            <artifactId>xz</artifactId>
          </exclusion>
        </exclusions>
      </dependency>
      <dependency>
        <groupId>org.apache.spark</groupId>
        <artifactId>spark-catalyst_${dependency.scala.version.suffix}</artifactId>
        <version>${dependency.spark.version}</version>
        <exclusions>
          <exclusion>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-log4j12</artifactId>
          </exclusion>
        </exclusions>
      </dependency>
      <!-- use found by dependency:analyze-only -->
      <dependency>
        <groupId>org.apache.spark</groupId>
        <artifactId>spark-network-common_${dependency.scala.version.suffix}</artifactId>
        <version>${dependency.spark.version}</version>
      </dependency>

      <dependency>
        <groupId>org.apache.spark</groupId>
        <artifactId>spark-sql_${dependency.scala.version.suffix}</artifactId>
        <version>${dependency.spark.version}</version>
        <exclusions>
          <exclusion>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-log4j12</artifactId>
          </exclusion>
        </exclusions>
      </dependency>
    </dependencies>
  </dependencyManagement>

  <dependencies>
    <dependency>
      <groupId>org.apache.spark</groupId>
      <artifactId>spark-catalyst_${dependency.scala.version.suffix}</artifactId>
    </dependency>

    <dependency>
      <groupId>org.apache.spark</groupId>
      <artifactId>spark-core_${dependency.scala.version.suffix}</artifactId>
    </dependency>

    <dependency>
      <groupId>org.apache.spark</groupId>
      <artifactId>spark-sql_${dependency.scala.version.suffix}</artifactId>
    </dependency>
  </dependencies>

  <build>
    <pluginManagement>
      <plugins>
        <!-- maven-pmd-plugin v 3.21.2 works with mvn clean pmd:pmd       -->
<!--        <plugin>-->
<!--          <groupId>org.apache.maven.plugins</groupId>-->
<!--          <artifactId>maven-pmd-plugin</artifactId>-->
<!--          <version>3.21.2</version>-->
<!--          <configuration>-->
<!--            <printFailingErrors>true</printFailingErrors>-->
<!--            <linkXRef>false</linkXRef>-->
<!--            <minimumTokens>100</minimumTokens>-->
<!--            <targetJdk>1.8</targetJdk>-->
<!--            <excludes>-->
<!--              <exclude>**/com/enveil/common/messages/*.java</exclude>-->
<!--              <exclude>**/gen/**/*.java</exclude>-->
<!--            </excludes>-->
<!--            <excludeRoots>-->
<!--              <excludeRoot>target/generated-sources</excludeRoot>-->
<!--              <excludeRoot>enveil-client/src/gen</excludeRoot>-->
<!--              <excludeRoot>src/gen</excludeRoot>-->
<!--              <excludeRoot>**/gen/**/*</excludeRoot>-->
<!--            </excludeRoots>-->
<!--          </configuration>-->
<!--          <executions>-->
<!--            <execution>-->
<!--              <goals>-->
<!--                <goal>pmd</goal>-->
<!--                <goal>check</goal>-->
<!--              </goals>-->
<!--              <phase>verify</phase>-->
<!--            </execution>-->
<!--          </executions>-->
<!--        </plugin>-->

        <!-- maven-pmd-plugin v 3.25.0 fails with mvn clean pmd:pmd       -->
        <plugin>
          <groupId>org.apache.maven.plugins</groupId>
          <artifactId>maven-pmd-plugin</artifactId>
          <version>3.25.0</version>
          <dependencies>
            <dependency>
              <groupId>net.sourceforge.pmd</groupId>
              <artifactId>pmd-core</artifactId>
              <version>${pmdVersion}</version>
            </dependency>
            <dependency>
              <groupId>net.sourceforge.pmd</groupId>
              <artifactId>pmd-java</artifactId>
              <version>${pmdVersion}</version>
            </dependency>
            <dependency>
              <groupId>net.sourceforge.pmd</groupId>
              <artifactId>pmd-javascript</artifactId>
              <version>${pmdVersion}</version>
            </dependency>
            <dependency>
              <groupId>net.sourceforge.pmd</groupId>
              <artifactId>pmd-jsp</artifactId>
              <version>${pmdVersion}</version>
            </dependency>
          </dependencies>
          <configuration>
            <printFailingErrors>true</printFailingErrors>
            <linkXRef>false</linkXRef>
            <minimumTokens>100</minimumTokens>
            <targetJdk>1.8</targetJdk>
            <excludes>
              <exclude>**/com/enveil/common/messages/*.java</exclude>
              <exclude>**/gen/**/*.java</exclude>
            </excludes>
            <excludeRoots>
              <excludeRoot>target/generated-sources</excludeRoot>
              <excludeRoot>enveil-client/src/gen</excludeRoot>
              <excludeRoot>src/gen</excludeRoot>
              <excludeRoot>**/gen/**/*</excludeRoot>
            </excludeRoots>
          </configuration>
          <executions>
            <execution>
              <goals>
                <goal>pmd</goal>
                <goal>check</goal>
              </goals>
              <phase>verify</phase>
            </execution>
          </executions>
        </plugin>
      </plugins>
    </pluginManagement>
  </build>
</project>
```

And in case its relevant. I'm using maven wrapper
```
wrapperVersion=3.3.2
distributionType=only-script
distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.9.8/apache-maven-3.9.8-bin.zip
```

**Comment by oowekyala** [comment]
From the error message it looks like the scala class `scala.PartialFunction$AndThen` has modifiers: public private static. This is a bug in ClassStub which I fixed in #5310, but this still needs tests
